### PR TITLE
openssl@3.0 3.0.8 (new formula)

### PR DESCRIPTION
**Pull request by ZhongRuoyu**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Resurrecting `openssl@3.0`, an LTS version. Currently 3.0 and 3.1 [do not differ much](https://www.openssl.org/docs/man3.1/man7/migration_guide.html#OPENSSL-3.1), but since  its EOL is [later than](https://www.openssl.org/policies/releasestrat.html) the current latest 3.x series, we may need this eventually.

Formula based on [`openssl@3.rb`](https://github.com/Homebrew/homebrew-core/blob/891026de513254b65ac322b87fe113f3510f355f/Formula/openssl@3.rb) before #125616.

### CODE REVIEW

**Comment by fxcoudert** [comment]
> Version 3.1 will be supported until 2025-03-14

We have time, and we do not know what will happen near the end of this cycle: will 3.1 be made LTS? will there be a 3.2 version with painless transition?

_“we may need this eventually”_ is a good argument to add it only when we actually need it :)

File: .github/workflows/triage.yml
Comment (MikeMcQuaid) [inline .github/workflows/triage.yml:164]: This shouldn't be necessary as we don't have lots of dependents

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good bar one nit!

**Comment by Bo98** [comment]
Yes, that's the reason we went for `@3` rather than `@3.0` initially.

**Comment by Porkepix** [comment]
> Yes, that's the reason we went for `@3` rather than `@3.0` initially.

So, what would be the need for one formula to rely on 3.0.X rather than 3.x, if it's 100% ABI/API compatible?

**Comment by ZhongRuoyu** [comment]
Thanks everyone for the review and comments! Several reasons why I'd like this to be added:

1. The direct reason being that I came across https://github.com/orgs/Homebrew/discussions/4344 and realised that the OpenSSL 3.0.x series is designated as the LTS. (I wouldn't have noticed this otherwise.)
2. As a consequence of being LTS, 3.0.8 is still listed on upstream's [download page](https://www.openssl.org/source/) despite the newer 3.1.0 release. So I *think* there should be community interest, more or less. So, I agree with @MikeMcQuaid in that "it makes us better and more useful for users with minimal overhead for us".
3. As @Porkepix and @Bo98 pointed out, 3.1.x releases (and as documented -- future 3.x series) ought to be ABI/API-compatible with the 3.0.x series. So, I don't expect the new `openssl@3.0`, after it is added, to have that many dependents -- most formulae can just continue using the latest `openssl@3`. Our https://github.com/Homebrew/homebrew-core/labels/openssl-3-migration shouldn't be affected by this either.

But I do understand and agree with @fxcoudert that this is currently not an essential need, at least from our (maintainers') perspective. Given the minor differences between the 3.1.0 and 3.0.x releases, if this has proven to be controversial, I'm not against postponing this. Maybe eventually it should be up to @Homebrew/tsc to decide.

**Comment by MikeMcQuaid** [comment]
> I don't care too much but I'd like there to at least be one reason before committing to adding this and supporting it for 3 years.

Adding a formula does not commit us to supporting it for 3 years.

One reason to add it: it's a useful package and, being a package manager, more packages make us more useful.

**Comment by fxcoudert** [comment]
> it's a useful package

It's 100% ABI compatible with the formula we ship, and we have never had a single user request. But I won't object further, @MikeMcQuaid's interest can count as a user request :)

File: .github/workflows/triage.yml
Comment (ZhongRuoyu) [inline .github/workflows/triage.yml:67]: I'm starting to reconsider if this should actually be declared as https://github.com/Homebrew/homebrew-core/labels/legacy. I mean, it's not really old, but `python@3.10` is not old either.

Comment (ZhongRuoyu) [reply .github/workflows/triage.yml]: Dropped this.

**Comment by Bo98** [comment]
> Adding a formula does not commit us to supporting it for 3 years.

Ok I guess 15 months then (3 months minimum deprecation, 1 year minimum disabled).

**Comment by MikeMcQuaid** [comment]
> > Adding a formula does not commit us to supporting it for 3 years.
> 
> Ok I guess 15 months then (3 months minimum deprecation, 1 year minimum disabled).*

No. This still does not commit us to anything. We can remove any formula at any time for any reason without deprecation or disabling. These are rules we've made for ourselves: we are not bound to any of them.

> Could do with more elaboration on the "useful" part given I don't see a use case over 3.1 given they're supposed to behave

If it literally never gets any installs after we add it for e.g. 90 days: I'll agree no-one finds it useful.

> The reason I ask is we don't add versioned formulae for every release ever made.

IMO we should be tending much more in this direction that we are. That we don't do this is a common gripe I hear about Homebrew.

> we can remove it outright without deprecation, though we've not really exercised that in a while (the LLVM & GHC formulae with 2 installs perhaps could do with a cleanup though).

Yup. Definitely game to remove these more aggressively. Could even (once we move to Influx) consider making this something we do automatically based on analytics data.

**Comment by pd0wm** [comment]
I don't want to hijack this discussion, but I had to downgrade to `3.0.8` due to issues with debugging a program linked against `3.1.0` using `lldb`. It immediately stops with `EXC_BAD_INSTRUCTION` on startup. Not sure if this is a fundamental issue with `3.1.X` or just a bug in the latest version.

<details>
  <summary>lldb output</summary>

```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=1, subcode=0x4a03000)
    frame #0: 0x0000000102063328 libcrypto.3.dylib`_armv8_sve_probe
libcrypto.3.dylib`:
->  0x102063328 <+0>: eor    z0.d, z0.d, z0.d
    0x10206332c <+4>: ret

libcrypto.3.dylib`:
    0x102063330 <+0>: xar    z0.d, z0.d, z0.d, #0x20
    0x102063334 <+4>: ret
```

</details>

**Comment by fxcoudert** [comment]
> I don't want to hijack this discussion, but I had to downgrade to 3.0.8 due to issues with debugging a program linked against 3.1.0 using lldb. It immediately stops with EXC_BAD_INSTRUCTION on startup. Not sure if this is a fundamental issue with 3.1.X or just a bug in the latest version.

The issue correspond to this commit: https://github.com/openssl/openssl/commit/a14eff6319ec254901e051d61e93d3272c451ebe
from this pull request: https://github.com/openssl/openssl/pull/17916

But it's a probe function, so it's expected that it gets EXC_BAD_INSTRUCTION: it's trying to see if SVE instructions are supported. In LLDB, you should let the exception propagate, it should be caught. And if you do not run in the debugger, you see no error, do you?

**Comment by pd0wm** [comment]
Noted! No errors without debugger. I'm probably not going to be the last person running into this 😅

**Comment by fxcoudert** [comment]
How about `process handle SIGILL --notify false --pass true --stop false` ?

**Comment by pd0wm** [comment]
> How about `process handle SIGILL --notify false --pass true --stop false` ?

As mentioned in my edit, that doesn't work. In the mono source code they mention the `EXC_BAD_ACCESS` exception is thrown first, and is translated into the POSIX signal by a host-wide exception handler. I'll do some more investigating, and take it up with the openssl team if needed. 

You can see this as a +1 vote from me for a separate `openssl@3.0` package 😄 .

**Comment by MikeMcQuaid** [comment]
@pd0wm @fxcoudert Let's move this into a separate issue, discussion please.



> You can see this as a +1 vote from me for a separate `openssl@3.0` package, at least until a new `3.1` release is made 😄 .

Yes, this was a useful contribution as a result, thanks @pd0wm!

**Review by MikeMcQuaid — APPROVED** [review]

**Review by chenrui333 — APPROVED** [review]

**Pull request merged**
